### DISCUSSION TITLE
Janet as a Bash script replacement: exec(3) family of functions?

### DISCUSSION THREAD

**Discussion by nrser** [janet-lang/janet#1303] [Q&A] [answered]
TL;DR: Is there support for [exec(3)][] functions somewhere I'm missing / in a package?

[exec(3)]: https://www.man7.org/linux/man-pages/man3/exec.3.html

Quick disclaimer: I've been looking at Janet for about one day, so my experience is pretty minimal.

I'm interested in Janet as a possible replacement for piles of Bash scripts. So lots of little programs that are run frequently, changed infrequently, and loosely owned. Fast start-up and highly regular syntax are the big selling points for Janet.

A common pattern in this role is to parse args, read some config, and assemble another command to swap in using the [exec(3)][] family of *nix/libc functions. Is there support for these functions somewhere in the stdlib/spork that I'm missing? If not, is it due to

1.  Inability to support a cross-platform API (unix-like only)
2. No one's really asked for / needed them
3. Just use package X ("exec" is a particularly difficult search term)
4. Something else...?

**Comment by tionis** [comment]
spork/argparse gives you argument parsing and os/execute allows you to execute commands.
I've also written a small helper for she'll scripts at http://tasadar.net/tionis/shell-tools.janet
And Andrewchamber's sh is also useful. Can write more later, my laptop is currently broken

**Comment by nstgc** [comment]
I'm not sure about a Janet equivalent to PISIX's `exec`, but I've found Janet to be an excellent replacement for Bash. I would recommend looking at https://janet.guide/scripting/ . Personally, I don't care for the recommended libraries (I prefer something more explicit/intentional feeling), but I'm sure a lot of people will and it is generally a very helpful guide (as are the other chapters).

**Reply by bakpakin** [reply, marked as answer]
Janet doesn't have fork and exec as separate functions, mostly for windows compatibility. That said, open a bug or PR and I would be happy to add them (on everything but windows).

For most needs, os/spawn is a wrapper around posix_spawn that has good analogues on windows.

**Reply by nrser** [reply]
Awesome, I'll open a bug. I did a quick native module implementation wrapping `execvp` just to see what that was like, but I'm new to Janet and my C is a bit rusty as well. It would be great to have support in the stdlib.

**Reply by nrser** [reply]
Issue is here: https://github.com/janet-lang/janet/issues/1304
